design/coroutines.md

```diff
@@ -0,0 +1,688 @@
+```
+
+and
+
+```c++
+wait(store(v, f));
+```
+
+becomes
+
+```c++
+v = co_await f;
+```
+
+### Replace Locals with Temporaries
+
+In certain places we use locals just to work around actor compiler limitations. Since locals use up space in the
+coroutine object they should be removed wherever it makes sense (only if it doesn't make the code less readable!).
+
+For example:
+
+```c++
+Foo f = wait(foo);
+bar(f);
+```
+
+might become
+
+```c++
+bar(co_await foo);
+```
+
+### Don't Wait in Error-Handlers
+
+Using `co_await` in an error-handler produces a compilation error in C++. However, this was legal with `ACTOR`. There
+is no general best way of addressing this issue, but usually it's quite easy to move the `co_await` expression out of
+the `catch`-block.
+
+One place where we use this pattern a lot if in our transaction retry loop:
+
+```c++
+state ReadYourWritesTransaction tr(db);
+loop {
+    try {
+        Value v = wait(tr.get(key));
+        tr.set(key2, val2);
+        wait(tr.commit());
+        return Void();
+    } catch (Error& e) {
+        wait(tr.onError(e));
+    }
+}
+```
+
+Luckily, with coroutines, we can do one better: generalize the retry loop. The above could look like this:
+
+```c++
+co_await db.run([&](ReadYourWritesTransaction* tr) -> Future<Void> {
+    Value v = wait(tr.get(key));
+    tr.set(key2, val2);
+    wait(tr.commit());
+});
+```
+
+A possible implementation of `Database::run` would be:
+
+```c++
+template <std:invocable<ReadYourWritesTransaction*> Fun>
+Future<Void> Database::run(Fun fun) {
+    ReadYourWritesTransaction tr(*this);
+    Future<Void> onError;
+    while (true) {
+        if (onError.isValid()) {
+            co_await onError;
+            onError = Future<Void>();
+        }
+        try {
+            co_await fun(&tr);
+        } catch (Error& e) {
+            onError = tr.onError(e);
+        }
+    }
+}
+```
+
+### Make Static Functions Class Members
+
+With actors, we often see the following pattern:
+
+```c++
+struct Foo : IFoo {
+    ACTOR static Future<Void> bar(Foo* self) {
+        // use `self` here to access members of `Foo`
+    }
+    
+    Future<Void> bar() override {
+        return bar(this);
+    }
+};
+```
+
+This boilerplate is necessary, because `ACTOR`s can't be class members since the actual code is moved into a different
+struct which.
+
+With C++ coroutines, this limitation goes away. So a cleaner (and slightly more efficient) implementation of the above
+is:
+
+```c++
+struct Foo : IFoo {
+    Future<Void> bar() override {
+        // `this` can be used like in any non-coroutine. `co_await` can be used.
+    }
+};
+```
+
+### Initialization of Locals
+
+There is one very subtle and hard to spot difference between `ACTOR` and a coroutine: the way some local variables are
+initialized. Consider the following code:
+
+```c++
+struct SomeStruct {
+    int a;
+    bool b;
+};
+
+ACTOR Future<Void> someActor() {
+    // beginning of body
+    state SomeStruct someStruct;
```

Comment:
Maybe worth mentioning that something like
```
state SomeStruct someStruct(1, false);
```
is invalid for flow, i.e., `SomeStruct` has to have the default constructor to be used here for the transpiled code. With the new co-routine, it becomes a valid one.

Reply:
I think it depends no? If your `state` variable is at the top of the function it will work as expected, otherwise it won't (which is arguably worse).

Reply:
Yes. I was assuming there are code before `state` statement.